tests/data_context/cloud_data_context/test_datasource_crud.py

```diff
@@ -15,52 +20,201 @@
+            "id": datasource_id,
+            "execution_engine": {
+                "class_name": "PandasExecutionEngine",
+                "module_name": "great_expectations.execution_engine",
+            },
+        }
+
+        return mock_response_factory(
+            {
+                "data": {
+                    "attributes": {"datasource_config": datasource_config_with_ids},
+                    "id": datasource_id,
+                }
+            },
+            200,
+        )
+
+    return _mocked_get_response
+
+
+class ConfigType(enum.Enum):
+    NAME_SUPPLIED_SEPARATELY: str = "name_supplied_separately"
+    CONFIG_INCLUDES_NAME: str = "config_includes_name"
+    NAME_SUPPLIED_SEPARATELY_AND_INCLUDED_IN_CONFIG: str = (
+        "name_supplied_separately_and_included_in_config"
+    )
+
+
+def _add_datasource_helper(
+    context: AbstractDataContext,
+    config_type: ConfigType,
+    save_changes: bool,
+    datasource_name: str,
+    datasource_config: DatasourceConfig,
+) -> BaseDatasource:
+    """Helper method to call add_datasource with the right parameters based on the test parametrization config."""
+    datasource_config_with_name: DatasourceConfig = copy.deepcopy(datasource_config)
+    datasource_config_with_name.name = datasource_name
+
+    # Call add_datasource with and without the name field included in the datasource config
+    stored_datasource: BaseDatasource
+    if isinstance(context, DataContext):
+        # Don't add the `save_changes` param for DataContext
+        name_and_save_changes_params_base: dict = {}
+    else:
+        name_and_save_changes_params_base = {"save_changes": save_changes}
+
+    name_and_save_changes_params: dict = name_and_save_changes_params_base.copy()
+    if config_type == ConfigType.NAME_SUPPLIED_SEPARATELY:
+        name_and_save_changes_params.update({"name": datasource_name})
+        datasource_config_dict: dict = datasource_config.to_dict().copy()
+        datasource_config_dict.pop(
+            "name", None
+        )  # make sure name (even if None) is not present
+        stored_datasource = context.add_datasource(
+            **name_and_save_changes_params,
+            **datasource_config_dict,
+        )
+    elif config_type == ConfigType.CONFIG_INCLUDES_NAME:
+        name_and_save_changes_params.update({})
+        stored_datasource = context.add_datasource(
+            **name_and_save_changes_params,
+            **datasource_config_with_name.to_dict(),
+        )
+    elif config_type == ConfigType.NAME_SUPPLIED_SEPARATELY_AND_INCLUDED_IN_CONFIG:
+        name_and_save_changes_params.update({"name": datasource_name})
+        stored_datasource = context.add_datasource(
+            **name_and_save_changes_params,
+            **datasource_config_with_name.to_dict(),
+        )
+    else:
+        raise Exception(
+            f"config_type must be one of {','.join([str(config_type) for config_type in ConfigType])}"
+        )
```

Comment:
Instead of branching logic like this in this helper, could we make 3 separate tests that cover each of these cases? I'm all for reducing code complexity but I'd rather we have duplication if it makes it easier to reason about the test.

Reply:
That's what we had before this PR (3 separate cases). The cases are super similar, the differences being how the name is included and the save changes setting. I'm thinking code re-use with this PR, and hopefully turn these into unit tests as much as possible as part of the testing effort we are undertaking now.

Reply:
Happy to close this PR though if you think this is a step backward.